### PR TITLE
Add tests for `Chain` implementation

### PR DESCRIPTION
**Pull request by yaahc**
## Related Issues

- https://github.com/ZcashFoundation/zebra/issues/1049 - tracking issue that describes the proptest design for this PR

## Related PRs

- https://github.com/ZcashFoundation/zebra/pull/1069 - the implementation this tests
- https://github.com/ZcashFoundation/zebra/pull/1092 - exports zebra-chain Arbitrary impls and strategies required to implement the Chain proptests

### CODE REVIEW

File: book/src/dev/rfcs/0005-state-updates.md [hunk 1]
```diff
@@ -283,14 +285,16 @@ struct Chain {
...
-    - Add new utxos and remove consumed utxos from `self.utxos`
-    - Add anchors to the appropriate `self.<version>_anchors`
-    - Add nullifiers to the appropriate `self.<version>_nullifiers`
+    - Add the block's hash to `height_by_hash`
     - Add work to `self.partial_cumulative_work`
+    - For each `transaction` in `block`
+      - Add key: `transaction.hash` and value: `(height, tx_index)` to `tx_by_hash`
+      - Add created utxos to `self.created_utxos`
+      - Add spent utxos to `self.spent_utxos`
+      - Add anchors to the appropriate `self.<version>_anchors`
```
Comment (dconnolly) [inline book/src/dev/rfcs/0005-state-updates.md]: This is not for each transaction, there will be one for Sprout and one for Sapling for each block, usually determined by last transaction(s) that have JoinSplit or Output descriptions, respectively, with exceptions for genesis and pre-Sapling blocks. These are passed in via the `Request::Commit(Finalized)Block` requests alongside the `Block` itself so maybe `Chain::push()` also needs those pre-computed values.

File: book/src/dev/rfcs/0005-state-updates.md [hunk 2]
```diff
@@ -283,14 +285,16 @@ struct Chain {
...
+      - Add nullifiers to the appropriate `self.<version>_nullifiers`
```
Comment (dconnolly) [inline book/src/dev/rfcs/0005-state-updates.md]: This is fine. 👍

File: book/src/dev/rfcs/0005-state-updates.md [hunk 3]
```diff
@@ -300,11 +304,13 @@ Remove the lowest height block of the non-finalized portion of a chain.
...
-      - remove `transaction.hash` from `tx_by_hash`
-    - Remove new utxos from `self.utxos`
-    - Remove the anchors from the appropriate `self.<version>_anchors`
-    - Remove the nullifiers from the appropriate `self.<version>_nullifiers`
+    - Subtract work from `self.partial_cumulative_work`
+    - For each `transaction` in `block`
+      - Remove `transaction.hash` from `tx_by_hash`
+      - Remove created utxos from `self.created_utxos`
+      - Remove spent utxos from `self.spent_utxos`
+      - Remove the anchors from the appropriate `self.<version>_anchors`
```
Comment (dconnolly) [inline book/src/dev/rfcs/0005-state-updates.md]: As noted above this is not really pre-transaction, there should be one each for Sprout and Sapling per block

File: zebra-state/src/memory_state.rs [hunk 4]
```diff
@@ -459,3 +459,179 @@ impl Ord for Chain {
...
+        {
+            (NoDebug(chain), count)
+        }
+    }
+
+    #[test]
+    fn forked_equals_pushed() -> Result<()> {
+        zebra_test::init();
+
+        proptest!(|((chain, count) in arbitrary_chain_and_count())| {
```
Comment (yaahc) [inline zebra-state/src/memory_state.rs]: as these are currently designed we will almost certainly want to disable these tests by default and add another CI step that builds the application in `--release` to run these tests

File: zebra-chain/src/block/arbitrary.rs [hunk 5]
```diff
@@ -67,7 +89,7 @@ impl Arbitrary for Header {
                     merkle_root: merkle_root_hash,
                     root_bytes,
                     time: Utc.timestamp(timestamp, 0),
-                    difficulty_threshold,
+                    difficulty_threshold: CompactDifficulty(545259519),
```
Comment (dconnolly) [inline zebra-chain/src/block/arbitrary.rs:93]: ?

File: zebra-chain/src/sapling/arbitrary.rs [hunk 6]
```diff
@@ -1,35 +1,33 @@
...
-                    spend_auth_sig: redjubjub::Signature::from({
-                        let mut b = [0u8; 64];
-                        b.copy_from_slice(sig_bytes.as_slice());
-                        b
-                    }),
-                },
-            )
+            .prop_map(|(anchor, nullifier, rpk_bytes, proof, sig_bytes)| Self {
+                anchor,
+                cv: ValueCommitment(AffinePoint::identity()),
```
Comment (dconnolly) [inline zebra-chain/src/sapling/arbitrary.rs:21]: 👍

File: zebra-chain/src/block/arbitrary.rs (hunk 5 above)
Comment (yaahc) [reply zebra-chain/src/block/arbitrary.rs:93]: ahahaha, one sec let me find where I calculated this

File: zebra-chain/src/sapling/arbitrary.rs [hunk 7]
```diff
@@ -41,23 +39,18 @@ impl Arbitrary for Output {
...
-                    ephemeral_key,
-                    enc_ciphertext,
-                    out_ciphertext,
-                    zkproof,
-                },
-            )
+            .prop_map(|(enc_ciphertext, out_ciphertext, zkproof)| Self {
+                cv: ValueCommitment(AffinePoint::identity()),
+                cm_u: NoteCommitment(AffinePoint::identity()).extract_u(),
+                ephemeral_key: keys::EphemeralPublicKey(AffinePoint::identity()),
```
Comment (dconnolly) [inline zebra-chain/src/sapling/arbitrary.rs:49]: 👍

File: zebra-chain/src/block/arbitrary.rs (hunk 5 above)
Comment (yaahc) [reply zebra-chain/src/block/arbitrary.rs:93]: it's this but evaluated so it doesn't have to export the constants, will be replaced once we can construct a CompactDifficulty from an Expanded difficulty

https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/work/difficulty/tests.rs#L148

File: zebra-state/src/memory_state.rs [hunk 8]
```diff
@@ -582,10 +589,47 @@ mod tests {
...
+
+            for block in chain {
+                full_chain.push(block);
+            }
+
+            for _ in 0..finalized_count {
+                let _finalized = full_chain.pop_root();
+            }
+
+            prop_assert_eq!(full_chain.blocks.len(), partial_chain.blocks.len());
```
Comment (dconnolly) [inline zebra-state/src/memory_state.rs:631]: <3 <3 <3

**Review by dconnolly — DISMISSED** [review]
i love tests

File: zebra-state/src/memory_state.rs (hunk 4 above)
Comment (teor2345) [reply zebra-state/src/memory_state.rs]: Let's fix the slow test issue before we merge?
Let's try to keep each test under 10 seconds.

I think it's ok to keep these tests on by default, if we also default to a small number of iterations.
Then we can build in release in CI, and set PROPTEST_CASES to something larger.

Here's how to make that change:
```rust
proptest! {
    // The block roundtrip test can be really slow, so we use fewer cases by
    // default. Set the PROPTEST_CASES env var to override this default.
    #![proptest_config(Config::with_cases(env::var("PROPTEST_CASES")
                                          .ok()
                                          .and_then(|v| v.parse().ok())
                                          .unwrap_or(16)))]
```
https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/block/tests/prop.rs#L48

File: zebra-chain/src/block/arbitrary.rs [hunk 9]
```diff
@@ -55,7 +89,7 @@ impl Arbitrary for Header {
...
```
Comment (teor2345) [inline zebra-chain/src/block/arbitrary.rs:93]:
```suggestion
                    // TODO: replace with `ExpandedDifficulty.to_compact` when that method is implemented
                    difficulty_threshold: CompactDifficulty(545259519),
```

File: zebra-chain/src/lib.rs [hunk 10]
```diff
@@ -6,7 +6,7 @@
 #![doc(html_favicon_url = "https://www.zfnd.org/images/zebra-favicon-128.png")]
 #![doc(html_logo_url = "https://www.zfnd.org/images/zebra-icon.png")]
 #![doc(html_root_url = "https://doc.zebra.zfnd.org/zebra_chain")]
-#![deny(missing_docs)]
+// #![deny(missing_docs)]
```
Comment (teor2345) [inline zebra-chain/src/lib.rs:9]: Let's make sure this TODO is on a tracking issue somewhere?

**Review by teor2345 — CHANGES_REQUESTED** [review]
Looks good!

Let's fix the slow tests issue before we merge, so we don't impact other developers.

Otherwise, this is all fantastic!

**Comment by teor2345** [comment]
@yaahc what do we need to do to get this merged?

**Comment by teor2345** [comment]
I merged #1090 into the state RFC, so we might have some conflicts here. Hopefully they will disappear in the rebase.

**Comment by yaahc** [comment]
> @yaahc what do we need to do to get this merged?

I think its just a bit of cleanup, make sure the tests dont run to wrong, etc. But first I want to try to see if I can't figure out why the sync is hanging when I use the full impl in the follow up PR

File: zebra-chain/src/lib.rs (hunk 10 above)
Comment (yaahc) [reply zebra-chain/src/lib.rs:9]: Went ahead and re enabled this in my follow up branch which should be the last PR in this chain

**Comment by teor2345** [comment]
> > @yaahc what do we need to do to get this merged?
> 
> I think its just a bit of cleanup, make sure the tests dont run to wrong, etc.

It would be nice to fix rustfmt before we merge.

> But first I want to try to see if I can't figure out why the sync is hanging when I use the full impl in the follow up PR

Can you point me to the specific PR?
And any relevant failure logs?
(Unfortunately, sync hangs can be hard to diagnose - we need better futures tooling to find livelocked and deadlocked futures.)

There are a lot of PRs open right now, so I want to make sure I have the right one.

File: zebra-state/src/memory_state.rs [hunk 11]
```diff
@@ -575,7 +575,11 @@ mod tests {
     fn forked_equals_pushed() -> Result<()> {
         zebra_test::init();
 
-        proptest!(|((chain, count) in arbitrary_chain_and_count())| {
+        proptest!(ProptestConfig::with_cases(env::var("PROPTEST_CASES")
+                                          .ok()
+                                          .and_then(|v| v.parse().ok())
+                                          .unwrap_or(1)),
```
Comment (teor2345) [inline zebra-state/src/memory_state.rs:581]: Oh my, 1 case.

Let's make a tracking item or ticket to look into proptest speed?

**Review by teor2345 — APPROVED** [review]
I think we're good here, as long as CI passes.

File: book/src/dev/rfcs/0005-state-updates.md (hunk 1 above)
Comment (yaahc) [reply book/src/dev/rfcs/0005-state-updates.md]: I'll update this in the followup PR that implements the chain_set and the service handling, based on our previous discussion i think it should be "add all the notecommitments in order" to some set

File: zebra-state/src/memory_state.rs (hunk 11 above)
Comment (yaahc) [reply zebra-state/src/memory_state.rs:581]: good idea. I have an idea for it already so I'll document that in the issue RN as I create it.

**Pull request merged**